src/main/scala/tech/cryptonomic/conseil/Harpoon.scala

```diff
@@ -0,0 +1,142 @@
+   * Described in cats-effect apis: https://typelevel.org/cats-effect/api/cats/effect/IOApp.html
+   */
+  override def run(args: List[String]): IO[ExitCode] = {
+
+    //we read custom conf, usually provided from the environment
+    val conf = HarpoonAppConfig
+
+    implicit val ctx = scala.concurrent.ExecutionContext.global
+    val timer = IO.timer(scala.concurrent.ExecutionContext.global)
+
+    //we create the processor stage, with the custom types needed
+    val tagging = AccountsTaggingProcess[IOState]
+
+    /* Implement any dependency needed to run the processing. */
+    implicit val serviceApis = injectServices(tagging)
+
+    /* now we can call the processing stage */
+    val processCycle = (processedTopLevel: tagging.ProcessorInput) =>
+      tagging.process
+        .run(processedTopLevel)
+        .flatTap {
+          case (level, flagged) =>
+            logger.pureLog[IO](
+              _.info(
+                "This Harpoon processing cycle marked {} accounts, up to level {}.",
+                flagged.fold("an unknown number of")(String.valueOf),
+                level.fold("unspecified")(String.valueOf)
+              )
+            )
+        }
+        .handleErrorWith(
+          err =>
+            logger.pureLog[IO](_.error("Harpoon processing failed to complete", err)) *>
+                IO(processedTopLevel, 0) //we reset to the previous state as to retry next round
+        )
+
+    //a wait period before running the process once more
+    val waitRound =
+      IO.shift *>
+          logger.pureLog[IO](
+            _.info("Harpoon is now waiting for the next processing cycle")
+          ) *>
+          timer.sleep(conf.cycleSleep)
+
+    //combine the two parts, recursively calling itself, with an updated state
+    def program(level: tagging.ProcessorInput): IO[Unit] =
+      for {
+        out <- processCycle(level)
+        (levelReached, flagCounts) = out
+        _ <- waitRound
+        _ <- program(levelReached)
+      } yield ()
+
+    /* we return the exit status into IO. The IOApp will run the actual process */
+    program(Option.empty)
+      .guarantee(
+        IO(dbHandle.close())
+      )
+      .as(ExitCode.Success)
+
+  }
+
+  /* Allows running db-operations as cats-effect IO actions */
+  private def runDbToIO[T] = IOUtil.runDbToIO[T](dbHandle.run _) _
+
+  /* Allows streaming db-results */
+  private def runDbToStream[R, T](action: StreamingDBIO[R, T]) =
+    IOUtil.publishStream(dbHandle.stream(action))
+
+  /* builds the service instance based on global operations available in the project */
+  private def injectServices(processor: AccountsTaggingProcess[IOState])(implicit ec: ExecutionContext) =
+    new processor.ServiceDependencies {
+
+      override def flagAsActive(accountId: TezosTypes.PublicKeyHash): IOState[Int] =
+        runDbToIO(TezosDatabaseOperations.flagAccount(accountId.value, AccountFlags.activated)).to[IOState]
```

Comment:
I'd suggest avoiding usage of `to`, as it only works on IO (and requires `cats.effect.Effect` otherwise), and using `State.liftF(...)` instead.

Reply:
Oh, you have that as `liftIOState` already. You can use that then.